### PR TITLE
[modbus] [WIP] Add support for Sungrow Hybrid Inverters

### PR DESCRIPTION
**Pull request by ferdisdot**
This PR adds support for Sungrow Hybrid Inverters connected via Modbus TCP.
I tested the binding with a Sungrow SH10RT device, but, acccording to  the register definition it should work for all devices of SH*RT series and for devices of SH*RS and SH*K* series.

As this is my first contribution to  openHAB I'm happy on any feedback.

### CODE REVIEW

File: bundles/org.openhab.binding.modbus.sungrow/src/main/java/org/openhab/binding/modbus/sungrow/internal/dto/InverterModelBlock13k.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.modbus.sungrow.internal.dto;
+
+import java.util.Optional;
+
+import org.apache.commons.lang3.builder.ReflectionToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sungrow/src/main/java/org/openhab/binding/modbus/sungrow/internal/dto/InverterModelBlock13k.java:18]: I believe you should rewrite the toString without apache commons. See https://github.com/openhab/openhab-addons/issues/7722. I am suggesting auto-generated methods with eclipse IDE.

File: bundles/org.openhab.binding.modbus.sungrow/src/main/java/org/openhab/binding/modbus/sungrow/internal/discovery/SungrowDiscoveryProcess.java [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        if (block.moduleID == FINAL_BLOCK) {
+            parsingFinished();
+        } else {
+            baseAddress += block.length;
+            if (block.moduleID == COMMON_BLOCK) {
+                readCommonBlock(block); // This is an asynchronous task
+                return;
+            } else {
+                // createDiscoveryResult(block);
+                // lookForSunSpecBlock();
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sungrow/src/main/java/org/openhab/binding/modbus/sungrow/internal/discovery/SungrowDiscoveryProcess.java:233]: Commented out code -- either remove or uncomment.

**Comment by fwolter** [comment]
@ferdisdot Did you address @ssalonen comments?

**Comment by ferdisdot** [comment]
Hi @fwolter, I started to work on it, but last month were quite busy. So the last few month I couldn't work on it. I hope I may get back to this until end of January. If you would like to contribute in the meantime feel free to do so.
As this is my first work in java I'm not sure about the type questions that @ssalonen raised (int vs. Integer). I would appreciate any help here.

**Comment by fwolter** [comment]
If you use `Integer` instead of `int` you're consuming a bit more resources and `Integer` can also be `null`, which needs to be handled correctly. So, using `int` makes things simpler and consumes less resources.

You made some changes to the code. Please raise your hand if you're read for another review round.

**Comment by lsiepel** [comment]
@ferdisdot what is the current state of this PR? Are there still things to address before another (hopefully last)  review round can take place?

**Comment by lsiepel** [comment]
Unfortunately it looks like we cannot bring this PR to the finish line. After two years of inactivity i close this PR. If there is a need to restart, feel free to re-open or start a new PR.

**Pull request closed without merging**